### PR TITLE
Rework cpu & thread base to support dynarmic

### PR DESCRIPTION
**Pull request by sunho**
# About 
1. CPU Abstraction
2. More readable, accurate, intuitive thread execution loop
3. Eliminate weird unintuitive callback flag in thread function
4. Use halt instruction to return control to host
5. More accurate context
6. Correctly implement float registers
7. Add basic dynarmic cpu implementation

### CODE REVIEW

File: vita3k/cpu/include/cpu/unicorn_cpu.h [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    void load_context(CPUContext context) override;
+
+    bool is_returning() override;
+    bool hit_breakpoint() override;
+    void trigger_breakpoint() override;
+    void set_log_code(bool log) override;
+    void set_log_mem(bool log) override;
+    bool get_log_code() override;
+    bool get_log_mem() override;
+};
```
Comment (Zangetsu38) [inline vita3k/cpu/include/cpu/unicorn_cpu.h:98]: No forget new line in end for all new file :)

File: vita3k/cpu/include/cpu/dynarmic_cpu.h [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    int step() override;
+
+    bool is_returning() override;
+    bool hit_breakpoint() override;
+    void trigger_breakpoint() override;
+    void set_log_code(bool log) override;
+    void set_log_mem(bool log) override;
+    bool get_log_code() override;
+    bool get_log_mem() override;
+};
```
Comment (Zangetsu38) [inline vita3k/cpu/include/cpu/dynarmic_cpu.h:68]: here too

File: vita3k/cpu/include/cpu/functions.h [hunk 3]
```diff
@@ -101,20 +79,12 @@ std::string disassemble(CPUState &state, uint64_t at, bool thumb, uint16_t *insn
...
+void set_log_mem(CPUState &state, bool log);
+bool get_log_code(CPUState &state);
+bool get_log_mem(CPUState &state);
 bool is_returning(CPUState &cpu);
 void set_thread_id(CPUState &state, SceUID thread_id);
 SceUID get_thread_id(CPUState &state);
 
-void push_lr(CPUState &cpu, Address lr);
-Address pop_lr(CPUState &cpu);
+std::unique_ptr<ModuleRegion> get_region(CPUState &state, Address addr);
```
Comment (Zangetsu38) [inline vita3k/cpu/include/cpu/functions.h:90]: here same

File: vita3k/interface.cpp [hunk 4]
```diff
@@ -573,4 +571,4 @@ ExitCode run_app(HostState &host, Ptr<const void> &entry_point) {
     }
 
     return Success;
-}
+}
```
Comment (Zangetsu38) [inline vita3k/interface.cpp:574]: here too

File: vita3k/cpu/src/dynarmic_cpu.cpp [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+    }
+
+    void ExceptionRaised(uint32_t pc, Dynarmic::A32::Exception exception) override {
+        switch (exception) {
+        case Dynarmic::A32::Exception::Breakpoint:
+            return;
+        case Dynarmic::A32::Exception::UndefinedInstruction: {
+            break;
+        }
+        default:
```
Comment (bookmist) [inline vita3k/cpu/src/dynarmic_cpu.cpp:136]: What is the difference? Return and break look different but do the same.

Comment (sunho) [reply vita3k/cpu/src/dynarmic_cpu.cpp:136]: Just remainder of print debugging. Thanks for point out.

File: vita3k/kernel/src/thread/thread.cpp [hunk 6]
```diff
@@ -280,24 +263,30 @@ int run_callback(KernelState &kernel, ThreadState &thread, const SceUID &thid, A
     // TODO: remaining arguments should be pushed into stack
     set_thread_id(*cpu, thid);
     write_pc(*cpu, callback_address);
-    CPUStatePtr old_thread_cpu = CPUStatePtr(cpu, [](CPUState *state) -> void {});
-    old_thread_cpu.swap(thread.cpu);
-    auto res = run(*cpu, true, callback_address);
-    thread.cpu.swap(old_thread_cpu);
+    write_lr(*cpu, cpu->halt_instruction_pc);
+    auto res = run(*cpu);
+
```
Comment (bookmist) [inline vita3k/kernel/src/thread/thread.cpp:268]: see https://github.com/Vita3K/Vita3K/commit/892f88cbff20d7439f00960d3ecd5ae67823f8ae
thread.cpu swap is important when callback is called from callback and both callbacks use stack. Check on atelier game

File: vita3k/modules/SceFiber/SceFiber.cpp [hunk 7]
```diff
@@ -29,27 +29,28 @@ InitialFiber *_findIntialFiber(KernelState &kernel, Address sp) {
...
         }
     }
     return nullptr;
 }
 
 void _resetFiber(HostState &host, SceFiber *fiber) {
-    auto ifiber = _findIntialFiber(host.kernel, fiber->cpu.sp);
-    assert(ifiber != nullptr);
-    memcpy(fiber, ifiber->fiber, sizeof(*fiber));
+    auto ifiber = _findIntialFiber(host.kernel, fiber->cpu.cpu_registers[13]);
```
Comment (bookmist) [inline vita3k/modules/SceFiber/SceFiber.cpp:40]: maybe constants for magic register numbers?

File: vita3k/modules/SceFiber/SceFiber.cpp [hunk 8]
```diff
@@ -29,27 +29,28 @@ InitialFiber *_findIntialFiber(KernelState &kernel, Address sp) {
...
 
 int _fiberSwitch(HostState &host, const ThreadStatePtr thread, SceFiber *fiber, CPUContext &backup_cpu_context, SceUInt32 argOnRunTo, Ptr<SceUInt32> argOnRun, bool reset) {
-    save_context(*(thread->cpu), backup_cpu_context);
+    backup_cpu_context = save_context(*thread->cpu);
 
     bool suspended = false;
     if (fiber->addrContext == 0 && reset) {
         _resetFiber(host, fiber);
-    } else if (fiber->entry.address() == fiber->cpu.pc) {
+    } else if (fiber->entry.address() == fiber->cpu.cpu_registers[15]) {
```
Comment (bookmist) [inline vita3k/modules/SceFiber/SceFiber.cpp:53]: maybe constants for magic register numbers?

File: vita3k/cpu/src/dynarmic_cpu.cpp [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+        static std::mutex mu;
+        mu.lock();
+        if (logged.find(addr) == logged.end()) {
+            LOG_TRACE("Fallback at addr 0x{:X}, inst 0x{:X} ({})", addr, MemoryReadCode(addr), disassemble(*parent, addr, nullptr));
+            logged.insert(addr);
+        }
+        mu.unlock();
+        CPUContext context = cpu->save_context();
+        cpu->fallback.load_context(context);
+        std::uint64_t res = static_cast<std::uint64_t>(cpu->fallback.execute_instructions_no_check(static_cast<int>(num_insts)));
```
[the remark this answers is not in the document]
Comment (sunho) [reply vita3k/cpu/src/dynarmic_cpu.cpp:122]: You're right. We must think about syscalls treading thread->cpu. Same for callback. If thread exit called from callback, it will not exit the thread because it stops callback cpu.

[the remark this answers is not in the document]
Comment (sunho) [reply vita3k/cpu/src/dynarmic_cpu.cpp:122]: Exitthread thing is little unambiguous though. It's very likely in vita os it just launch callback on another thread.

[the remark this answers is not in the document]
Comment (sunho) [reply vita3k/cpu/src/dynarmic_cpu.cpp:122]: btw, I just noticed that there's no way fallback will happen on systemcall. Because svc is implemented. Anyway, it's good to consider.

**Comment by Meowlithius** [comment]
This definitely needs to get done otherwise Unicorn is slowing this project down massively. This probably needs a few labels attached too. such as _Refactoring_, _CPU_, _Discussion_, and a new _Project Priority_ Label.

File: vita3k/interface.cpp [hunk 10]
```diff
@@ -573,4 +573,4 @@ ExitCode run_app(HostState &host, Ptr<const void> &entry_point) {
...
```
Comment (Zangetsu38) [inline vita3k/interface.cpp:576]: hmm you can revert this change file ^^

File: .gitmodules [hunk 11]
```diff
@@ -84,3 +84,9 @@
 [submodule "external/xxHash"]
 	path = external/xxHash
 	url = https://github.com/Cyan4973/xxHash
+[submodule "external/dynarmic"]
+	path = external/dynarmic
+url=https://github.com/Vita3K/dynarmic.git
```
Comment (scribam) [inline .gitmodules:89]: The line format looks wrong to me.
Also, why a fork? Is it really necessary?

Comment (sunho) [reply .gitmodules:89]: Fork is very likely to be necessary. Not sure about formatting. Should I just manually format this?

Comment (Zangetsu38) [reply .gitmodules:89]: i have see and already fixed it in poroper branch here
https://github.com/Zangetsu38/Vita3K/commits/dynarmic
i can force push in your branch if yoiu are right :)

Comment (scribam) [reply .gitmodules:89]: Manual formatting is fine.
I understand a fork is the quicker way to add your changes but wouldn't it better to add them upstream and have feedbacks from dynarmic's maintainers?

**Review by Zangetsu38 — APPROVED** [review]

File: .gitmodules (hunk 11 above)
Comment (sunho) [reply .gitmodules]: We'll definitely pr to upstream. But, feedbacks can be little slow when Mary is busy. I think having a fork can help us experiment with changes in vita3k early on.
